contracts/src/L1/L1MigrationController.sol

```diff
@@ -0,0 +1,143 @@
+     * Implements ERC1155Receiver.onERC1155Received
+     */
+    function onERC1155Received(address /*operator*/, address /*from*/, uint256 tokenId, uint256 /*amount*/, bytes calldata data) external virtual returns (bytes4) {
+        if (msg.sender != address(nameWrapper)) {
+            revert UnauthorizedCaller(msg.sender);
+        }
+
+        (MigrationData memory migrationData) = abi.decode(data, (MigrationData));
+        MigrationData[] memory migrationDataArray = new MigrationData[](1);
+        migrationDataArray[0] = migrationData;
+
+        uint256[] memory tokenIds = new uint256[](1);
+        tokenIds[0] = tokenId;
+
+        _migrateWrappedEthNames(tokenIds, migrationDataArray);
+        
+        return this.onERC1155Received.selector;
+    }
+
+    /**
+     * Implements ERC1155Receiver.onERC1155BatchReceived
+     */
+    function onERC1155BatchReceived(address /*operator*/, address /*from*/, uint256[] memory tokenIds, uint256[] memory /*amounts*/, bytes calldata data) external virtual returns (bytes4) {
+        if (msg.sender != address(nameWrapper)) {
+            revert UnauthorizedCaller(msg.sender);
+        }
+
+        (MigrationData[] memory migrationDataArray) = abi.decode(data, (MigrationData[]));
+
+        _migrateWrappedEthNames(tokenIds, migrationDataArray);
+
+        return this.onERC1155BatchReceived.selector;
+    }
+
+    /**
+     * @dev Implements ERC721Receiver.onERC721Received
+     *
+     * If this is called then it means an unwrapped .eth name is being migrated to v2.
+     */
+    function onERC721Received(address /*operator*/, address /*from*/, uint256 tokenId, bytes calldata data) external virtual returns (bytes4) {
+        if (msg.sender != address(ethRegistryV1)) {
+            revert UnauthorizedCaller(msg.sender);
+        }
+        
+        (MigrationData memory migrationData) = abi.decode(data, (MigrationData));
+
+        _migrateNameViaBridge(tokenId, migrationData);
+
+        return this.onERC721Received.selector;
+    }
+
+    // Internal functions
+
+    /**
+     * @dev Called when wrapped .eth names are being migrated to v2.
+     * Checks if names are locked (have CANNOT_UNWRAP burned) and routes accordingly.
+     *
+     * @param tokenIds The token IDs of the .eth names.
+     * @param migrationDataArray The migration data for each .eth name.
+     */
+    function _migrateWrappedEthNames(uint256[] memory tokenIds, MigrationData[] memory migrationDataArray) internal {                
+        for (uint256 i = 0; i < tokenIds.length; i++) {
+            (, uint32 fuses, ) = nameWrapper.getData(tokenIds[i]);
+            
+            if (fuses & CANNOT_UNWRAP != 0) { // Name is locked
+                if (address(strategy) == address(0)) {
+                    revert NoMigrationStrategySet();
+                }
+                // Name is locked, migrate through strategy
+                strategy.migrateLockedEthName(address(nameWrapper), tokenIds[i], migrationDataArray[i]);
```

Comment:
We need to think about how we'll handle migrating subdomains, if the strategy supports it, too.

Reply:
Yeah, I was going to leave this to be handled within the concerete strategy